File: packages/agent-core/src/tools/builtin/code-quality/harvest-ody-markers.ts
import { z } from 'zod';
import type { Kaos } from '@odysseythink/kaos';
import type { BuiltinTool } from '#/agent/tool';
import type { ExecutableToolContext, ExecutableToolResult, RunnableToolExecution, ToolExecution } from '#/loop/types';
import type { TelemetryClient } from '#/telemetry';
import { resolvePathAccessPath } from '#/tools/policies/path-access';
import { toInputJsonSchema } from '@odysseythink/agent-core-shared';
import { literalRulePattern } from '#/tools/support/rule-match';
import type { WorkspaceConfig } from '#/tools/support/workspace';
import { GrepInputSchema, GrepTool } from '#/tools/builtin/file/grep';

export const MAX_MARKERS = 200;

// ---- Data models ----

export interface DebtLedgerMarker {
  readonly file: string;
  readonly line: number;
  readonly ceiling: string;
  readonly upgrade: string;
  readonly rot: boolean;
}

export const HarvestOdyMarkersInputSchema = z.object({
  path: z.string().optional().describe('Optional subdirectory or file to scan.'),
});

export type HarvestOdyMarkersInput = z.infer<typeof HarvestOdyMarkersInputSchema>;

export interface HarvestOdyMarkersOutput {
  readonly markdown: string;
  readonly markerCount: number;
  readonly rotRiskCount: number;
  readonly truncated: boolean;
}

// ---- Parser ----

const MARKER_RE =
  /^(.+?):(\d+):\s*(?:(?:\/\/)|#)\s*ody:\s*(.*)$/;

export function parseOdyMarker(rawLine: string): DebtLedgerMarker | null {
  const match = MARKER_RE.exec(rawLine);
  if (!match) return null;

  const file = match[1]!;
  const line = Number(match[2]);
  const body = match[3]!.trim();

  const commaIdx = body.indexOf(',');
  if (commaIdx === -1) {
    return { file, line, ceiling: body, upgrade: '', rot: true };
  }

  const ceiling = body.slice(0, commaIdx).trim();
  const upgrade = body.slice(commaIdx + 1).trim();
  const rot = upgrade === '';
  return { file, line, ceiling, upgrade, rot };
}

// ---- Renderer ----

export function renderDebtLedger(
  markers: readonly DebtLedgerMarker[],
  truncated: boolean,
): string {
  if (markers.length === 0) {
    return '未找到 `ody:` 债务标记。台账干净。';
  }

  const groups = new Map<string, DebtLedgerMarker[]>();
  for (const m of markers) {
    const list = groups.get(m.file);
    if (list) {
      list.push(m);
    } else {
      groups.set(m.file, [m]);
    }
  }

  const sortedFiles = [...groups.keys()].sort();

  const lines: string[] = [];
  for (const file of sortedFiles) {
    lines.push(`### ${file}`);
    const fileMarkers = groups.get(file)!;
    fileMarkers.sort((a, b) => a.line - b.line);
    for (const m of fileMarkers) {
      const rotTag = m.rot ? ' ⚠️ rot' : '';
      const upgradeDisplay = m.upgrade || '（未指定）';
      lines.push(
        `${m.file}:${m.line} — ${m.ceiling}。天花板：${m.ceiling}。升级：${upgradeDisplay}${rotTag}`,
      );
    }
    lines.push('');
  }

  const totalRot = markers.filter((m) => m.rot).length;
  lines.push(
    `**汇总**：${markers.length} 个标记，${totalRot} 个 rot 风险。`,
  );
  if (truncated) {
    lines.push('结果已截断至前 200 条；如需完整扫描，请指定更小的目录或文件。');
  }
  return lines.join('\n');
}

// ---- Tool ----

const TOOL_DESCRIPTION =
  'Scan the codebase for `// ody:` / `# ody:` simplification-debt markers and return a Chinese-first ledger report.';

export class HarvestOdyMarkersTool
  implements BuiltinTool<HarvestOdyMarkersInput>
{
  readonly name = 'HarvestOdyMarkers' as const;
  readonly description = TOOL_DESCRIPTION;
  readonly parameters: Record<string, unknown> =
    toInputJsonSchema(HarvestOdyMarkersInputSchema);

  constructor(
    private readonly kaos: Kaos,
    private readonly workspace: WorkspaceConfig,
    private readonly grepTool: GrepTool,
    private readonly telemetry: TelemetryClient,
  ) {}

  resolveExecution(args: HarvestOdyMarkersInput): ToolExecution {
    return {
      description: 'Harvesting ody: simplification debt markers',
      approvalRule: literalRulePattern(this.name, ''),
      execute: ({ signal }) => this.run(args, signal),
    };
  }

  private async run(
    args: HarvestOdyMarkersInput,
    signal: AbortSignal,
  ): Promise<ExecutableToolResult> {
    try {
      const scanPath = args.path ?? this.workspace.workspaceDir;

      const resolvedPath = resolvePathAccessPath(scanPath, {
        kaos: this.kaos,
        workspace: this.workspace,
        operation: 'search',
        policy: {
          guardMode: 'absolute-outside-allowed',
          checkSensitive: false,
        },
      });

      const grepInput = GrepInputSchema.parse({
        pattern: '(#|//) ?ody:',
        path: resolvedPath,
        output_mode: 'content',
        '-n': true,
        head_limit: MAX_MARKERS,
        include_ignored: false,
      });

      const grepExecution = this.grepTool.resolveExecution(grepInput);
      const grepCtx: ExecutableToolContext = {
        turnId: 'internal',
        toolCallId: 'internal',
        signal,
      };
      const grepResult = await (grepExecution as RunnableToolExecution).execute(grepCtx);

      if (grepResult.isError) {
        this.telemetry.track('debt_ledger_failed', {
          error: String(grepResult.output),
        });
        return {
          isError: true,
          output: `债务台账扫描失败：${String(grepResult.output)}`,
        };
      }

      const outputText = String(grepResult.output);
      const rawLines = outputText
        .split('\n')
        .filter((l: string) => l.trim() !== '');
      const markers = rawLines
        .map((line: string) => parseOdyMarker(line))
        .filter((m: DebtLedgerMarker | null): m is DebtLedgerMarker => m !== null);

      const truncated = outputText.includes('Results truncated');

      const markdown = renderDebtLedger(markers, truncated);

      this.telemetry.track('debt_ledger_harvested', {
        marker_count: markers.length,
        rot_risk_count: markers.filter((m: DebtLedgerMarker) => m.rot).length,
      });

      const output: HarvestOdyMarkersOutput = {
        markdown,
        markerCount: markers.length,
        rotRiskCount: markers.filter((m: DebtLedgerMarker) => m.rot).length,
        truncated,
      };

      return { isError: false, output: JSON.stringify(output) };
    } catch (error) {
      const message =
        error instanceof Error ? error.message : String(error);
      this.telemetry.track('debt_ledger_failed', { error: message });
      return {
        isError: true,
        output: `债务台账扫描失败：${message}`,
      };
    }
  }
}

File: rust-ody/crates/tools-rs/src/builtin/quality/harvest_ody_markers.rs
use std::sync::Arc;

use kaos_rs::environment::Environment;
use kaos_rs::kaos::Kaos;
use serde_json::Value;

use crate::builtin::grep::GrepTool;
use crate::builtin::quality::{parse_ody_marker, render_debt_ledger, NoopTelemetryClient};
use crate::builtin::{
    BuiltinTool, ExecutableToolContext, ExecutableToolResult, ToolError, ToolExecution,
};
use crate::policies::path_access::{
    assert_path_allowed, AssertPathOptions, PathAccessOperation, PathClass,
};
use crate::policies::rule_match::literal_rule_pattern;
use crate::tool_accesses::ToolAccesses;
use crate::workspace::WorkspaceConfig;

const MAX_MARKERS: i64 = 200;
const TOOL_DESCRIPTION: &str = "Scan the codebase for `// ody:` / `# ody:` simplification-debt markers and return a Chinese-first ledger report.";

pub struct HarvestOdyMarkersTool {
    kaos: Kaos,
    workspace: WorkspaceConfig,
    #[allow(dead_code)]
    grep: GrepTool,
    telemetry: Arc<dyn crate::builtin::quality::TelemetryClient>,
}

impl HarvestOdyMarkersTool {
    pub fn new(
        kaos: Kaos,
        workspace: WorkspaceConfig,
        grep: GrepTool,
        telemetry: Arc<dyn crate::builtin::quality::TelemetryClient>,
    ) -> Self {
        Self {
            kaos,
            workspace,
            grep,
            telemetry,
        }
    }
}

impl BuiltinTool for HarvestOdyMarkersTool {
    fn name(&self) -> &str {
        "HarvestOdyMarkers"
    }

    fn description(&self) -> &str {
        TOOL_DESCRIPTION
    }

    fn parameters(&self) -> Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "Optional subdirectory or file to scan."
                }
            },
            "additionalProperties": false
        })
    }

    fn resolve_execution(&self, args: Value) -> Result<ToolExecution, ToolError> {
        let path_arg = args
            .get("path")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();
        let scan_path = if path_arg.is_empty() {
            self.workspace.workspace_dir.clone()
        } else {
            path_arg
        };

        let path_class = match self.kaos.path_class() {
            "win32" => PathClass::Win32,
            _ => PathClass::Posix,
        };

        let safe_path = assert_path_allowed(
            &scan_path,
            &self.kaos.getcwd(),
            &self.workspace,
            AssertPathOptions {
                mode: PathAccessOperation::Search,
                check_sensitive: Some(false),
                path_class: Some(path_class),
            },
        )?;

        let approval_rule = literal_rule_pattern(self.name(), &safe_path);
        let kaos = self.kaos.clone();
        let workspace = self.workspace.clone();
        let telemetry = Arc::clone(&self.telemetry);
        let safe_path2 = safe_path.clone();

        Ok(ToolExecution {
            accesses: ToolAccesses::search_tree(&safe_path),
            description: "Harvesting ody: simplification debt markers".into(),
            approval_rule,
            matches_rule: None,
            display: None,
            execute: Box::new(move |ctx| {
                let safe_path = safe_path2.clone();
                let workspace_dir = workspace.workspace_dir.clone();
                let grep = GrepTool::new(kaos.clone(), workspace.clone());
                let telemetry = Arc::clone(&telemetry);
                Box::pin(async move {
                    run_harvest(safe_path, workspace_dir, grep, telemetry, ctx).await
                })
            }),
        })
    }
}

async fn run_harvest(
    safe_path: String,
    workspace_dir: String,
    grep: GrepTool,
    telemetry: Arc<dyn crate::builtin::quality::TelemetryClient>,
    ctx: ExecutableToolContext,
) -> ExecutableToolResult {
    if ctx.signal.aborted() {
        return ExecutableToolResult::error_text("Aborted".into(), "Aborted".into());
    }

    let grep_args = serde_json::json!({
        "pattern": "(#|//) ?ody:",
        "path": safe_path,
        "output_mode": "content",
        "-n": true,
        "head_limit": MAX_MARKERS,
        "include_ignored": false,
    });

    let exec = match grep.resolve_execution(grep_args) {
        Ok(e) => e,
        Err(e) => {
            telemetry.track(
                "debt_ledger_failed",
                serde_json::json!({ "error": e.to_string() }),
            );
            return ExecutableToolResult::error_text(
                format!("债务台账扫描失败：{}", e),
                "Grep setup failed".into(),
            );
        }
    };

    let grep_result = (exec.execute)(ctx).await;

    if grep_result.is_error {
        let err_text = grep_result.to_text();
        telemetry.track(
            "debt_ledger_failed",
            serde_json::json!({ "error": err_text }),
        );
        return ExecutableToolResult::error_text(
            format!("债务台账扫描失败：{}", err_text),
            "Grep failed".into(),
        );
    }

    let output_text = grep_result.to_text();
    let raw_lines: Vec<&str> = output_text
        .lines()
        .filter(|l| !l.trim().is_empty())
        .collect();
    let mut markers: Vec<crate::builtin::quality::DebtLedgerMarker> =
        raw_lines.into_iter().filter_map(parse_ody_marker).collect();
    for m in &mut markers {
        m.file = relativize_to_workspace(&m.file, &workspace_dir);
    }

    let truncated = output_text.contains("Results truncated");
    let markdown = render_debt_ledger(&markers, truncated);
    let rot_count = markers.iter().filter(|m| m.rot).count();

    telemetry.track(
        "debt_ledger_harvested",
        serde_json::json!({
            "marker_count": markers.len(),
            "rot_risk_count": rot_count,
        }),
    );

    let payload = serde_json::json!({
        "markdown": markdown,
        "markerCount": markers.len(),
        "rotRiskCount": rot_count,
        "truncated": truncated,
    });

    ExecutableToolResult::ok_text(payload.to_string())
}

fn relativize_to_workspace(file: &str, workspace_dir: &str) -> String {
    if workspace_dir.is_empty() {
        return file.to_string();
    }
    if file == workspace_dir {
        return ".".to_string();
    }
    let prefix = format!("{}/", workspace_dir);
    if let Some(rel) = file.strip_prefix(&prefix) {
        rel.to_string()
    } else {
        file.to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn dummy_env() -> Environment {
        Environment {
            os_kind: "macOS".to_string(),
            os_arch: "arm64".to_string(),
            os_version: "23.0.0".to_string(),
            shell_name: "bash".to_string(),
            shell_path: "/bin/bash".to_string(),
        }
    }

    #[tokio::test]
    async fn harvests_markers_from_temp_tree() {
        let tmp = tempfile::tempdir().unwrap();
        let kaos = Kaos::new(dummy_env(), tmp.path());
        let ws = WorkspaceConfig::new(tmp.path().to_string_lossy().to_string());

        // Write a file with two markers.
        std::fs::write(
            tmp.path().join("a.rs"),
            "fn main() {\n  // ody: hardcoded timeout, use config\n  // ody: missing test\n}\n",
        )
        .unwrap();

        let grep = GrepTool::new(kaos.clone(), ws.clone());
        let tool = HarvestOdyMarkersTool::new(kaos, ws, grep, Arc::new(NoopTelemetryClient));
        let exec = tool.resolve_execution(serde_json::json!({})).unwrap();
        let result = (exec.execute)(ExecutableToolContext {
            turn_id: "1".into(),
            tool_call_id: "call_1".into(),
            signal: crate::builtin::AbortSignal::new(),
            metadata: None,
        })
        .await;

        assert!(!result.is_error, "unexpected error: {}", result.to_text());
        let text = result.to_text();
        assert!(
            text.contains("hardcoded timeout"),
            "missing ceiling: {}",
            text
        );
        assert!(
            text.contains("missing test"),
            "missing second marker: {}",
            text
        );
        assert!(text.contains("2 个标记"), "missing count: {}", text);
    }

    #[tokio::test]
    async fn empty_tree_returns_clean_ledger() {
        let tmp = tempfile::tempdir().unwrap();
        let kaos = Kaos::new(dummy_env(), tmp.path());
        let ws = WorkspaceConfig::new(tmp.path().to_string_lossy().to_string());
        let grep = GrepTool::new(kaos.clone(), ws.clone());
        let tool = HarvestOdyMarkersTool::new(kaos, ws, grep, Arc::new(NoopTelemetryClient));
        let exec = tool.resolve_execution(serde_json::json!({})).unwrap();
        let result = (exec.execute)(ExecutableToolContext {
            turn_id: "1".into(),
            tool_call_id: "call_1".into(),
            signal: crate::builtin::AbortSignal::new(),
            metadata: None,
        })
        .await;

        assert!(!result.is_error);
        assert!(result.to_text().contains("台账干净"));
    }
}
